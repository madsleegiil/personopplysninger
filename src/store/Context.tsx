import React, { createContext, useContext, useReducer, Dispatch } from "react";
import { Action, Locale, reducer, Store } from "./Store";
import { FetchNameInfo } from "./providers/Auth";
import { FetchFeatureToggles } from "./providers/FeatureToggles";
import { FetchDsopInfo } from "../pages/digital-samhandling-offentlig-privat/DsopFetch";
import { FetchInstInfo } from "../pages/institusjonsopphold/InstFetch";
import { FetchPersonInfo } from "./providers/PersonInfo";
import { FetchKontaktInfo } from "../pages/forside/sections/4-personinfo/2-kontaktinfo/subsections/kontakt-og-reservasjonsregisteret/DKIF-Fetch";
import { Fetchskattetrekksmeldinger } from "../pages/skattetrekksmelding/SkattFetch";
import { FetchMedlInfo } from "../pages/medlemskap-i-folketrygden/MedlFetch";

interface Props {
  children: JSX.Element | JSX.Element[];
}

export const StoreContext = createContext({} as [Store, Dispatch<Action>]);
export const StoreProvider = (props: Props) => {
  const { children } = props;

  const initialLocale = (window.location.pathname.includes("/en")
    ? "en"
    : "nb") as Locale;

  const initialState = {
    locale: initialLocale,
    nameInfo: { status: "LOADING" } as FetchNameInfo,
    featureToggles: {
      status: "LOADING",
      data: {
        "personopplysninger.pdl": false,
        "personopplysninger.dsop": false,
        "personopplysninger.inst": false,
        "personopplysninger.skatt": false,
        "personopplysninger.medl": false,
        "personopplysninger.fullmakt": false,
        "personopplysninger.tilrettelegging": false,
        "pdl-fullmakt": false,
      },
    } as FetchFeatureToggles,
    dsopInfo: { status: "LOADING" } as FetchDsopInfo,
    instInfo: { status: "LOADING" } as FetchInstInfo,
    personInfo: { status: "LOADING" } as FetchPersonInfo,
    kontaktInfo: { status: "LOADING" } as FetchKontaktInfo,
    skattetrekksmeldinger: { status: "LOADING" } as Fetchskattetrekksmeldinger,
    medlInfo: { status: "LOADING" } as FetchMedlInfo,
  };

  return (
    <StoreContext.Provider value={useReducer(reducer, initialState)}>
      {children}
    </StoreContext.Provider>
  );
};
export const useStore = () => useContext(StoreContext);
