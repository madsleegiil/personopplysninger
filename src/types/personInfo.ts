import { Personalia } from "./personalia";
import { Adresser } from "./adresser";

export interface PersonInfo {
  personalia?: Personalia;
  adresser?: Adresser;
}
