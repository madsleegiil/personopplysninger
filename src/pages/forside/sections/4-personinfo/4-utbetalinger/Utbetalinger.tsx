import React, { useState } from "react";
import Box from "components/box/Box";
import kontonummerIkon from "assets/img/Kontonummer.svg";
import { UtenlandskBankkonto } from "types/personalia";
import Kilde from "components/kilde/Kilde";
import endreIkon from "assets/img/Pencil.svg";
import leggTilIkon from "assets/img/Pencil.svg";
import NorskKontonummer from "./visning/NorskKontonummer";
import Utenlandskonto from "./visning/UtenlandsBankkonto";
import OpprettEllerEndreNorskKontonr from "./endring/NorskKontonummer";
import { setOutboundNorskKontonummer } from "./endring/NorskKontonummer";
import { OutboundNorskKontonummer } from "./endring/NorskKontonummer";
import OpprettEllerEndreUtenlandsbank from "./endring/utenlandsk-bankkonto/UtenlandsBankkonto";
import { setOutboundUtenlandsbankonto } from "./endring/utenlandsk-bankkonto/UtenlandsBankkonto";
import { OutboundUtenlandsbankonto } from "./endring/utenlandsk-bankkonto/UtenlandsBankkonto";
import { Radio, RadioGruppe } from "nav-frontend-skjema";
import { FormattedMessage, useIntl } from "react-intl";
import { Knapp } from "nav-frontend-knapper";
import Alert, { AlertType } from "components/alert/Alert";
import { Form, FormContext, Validation } from "calidation";
import { fetchPersonInfo, postKontonummer } from "clients/apiClient";
import { PersonInfo } from "types/personInfo";
import { useStore } from "store/Context";
import driftsmeldinger from "driftsmeldinger";
import { AlertStripeAdvarsel } from "nav-frontend-alertstriper";

interface Props {
  utenlandskbank?: UtenlandskBankkonto;
  personident?: { verdi: string; type: string };
  kontonr?: string;
}

const NORSK = "NORSK";
const UTENLANDSK = "UTENLANDSK";

const Utbetalinger = (props: Props) => {
  const { formatMessage: msg } = useIntl();
  const { kontonr, utenlandskbank, personident } = props;
  const [loading, settLoading] = useState<boolean>(false);
  const [opprettEllerEndre, settOpprettEllerEndre] = useState<boolean>();
  const [alert, settAlert] = useState<AlertType | undefined>();
  const [, dispatch] = useStore();

  const initialValues = {
    norskEllerUtenlandsk: kontonr
      ? NORSK
      : utenlandskbank
      ? UTENLANDSK
      : undefined,
  };

  const config = {
    norskEllerUtenlandsk: {
      isRequired: msg({ id: "felter.type.velg" }),
    },
  };

  const submitEndre = (context: FormContext) => {
    const { isValid, fields } = context;
    if (isValid) {
      type Outbound = OutboundNorskKontonummer | OutboundUtenlandsbankonto;
      const outbound: { [key: string]: () => Outbound } = {
        NORSK: () => setOutboundNorskKontonummer(context),
        UTENLANDSK: () => setOutboundUtenlandsbankonto(context),
      };

      settLoading(true);
      postKontonummer(outbound[fields.norskEllerUtenlandsk]())
        .then(getUpdatedData)
        .then(onSuccess)
        .catch((error: AlertType) => settAlert(error))
        .then(() => settLoading(false));
    }
  };

  const getUpdatedData = () =>
    fetchPersonInfo().then((personInfo) => {
      dispatch({
        type: "SETT_PERSON_INFO_RESULT",
        payload: personInfo as PersonInfo,
      });
    });

  const onSuccess = () => {
    settOpprettEllerEndre(false);
  };

  return (
    <Box id="utbetaling" tittel="utbetalinger.tittel" icon={kontonummerIkon}>
      <>
        {driftsmeldinger.pdl && (
          <div style={{ paddingBottom: "1rem" }}>
            <AlertStripeAdvarsel>{driftsmeldinger.pdl}</AlertStripeAdvarsel>
          </div>
        )}
        <Alert type={"feil"} code={0} text={"Vi har for tiden tekniske problemer med å vise utenlandske kontonumre. Vi jobber med å rette feilen."} />
      </>
      {opprettEllerEndre ? (
        <Form onSubmit={submitEndre}>
          <Validation config={config} initialValues={initialValues}>
            {({ submitted, isValid, errors, setField, fields }) => {
              return (
                <RadioGruppe feil={submitted && errors.norskEllerUtenlandsk}>
                  <Radio
                    name={NORSK}
                    checked={fields.norskEllerUtenlandsk === NORSK}
                    label={msg({ id: "felter.kontonummervalg.norsk" })}
                    onChange={(e) =>
                      setField({ norskEllerUtenlandsk: e.target.name })
                    }
                  />
                  {fields.norskEllerUtenlandsk === NORSK && (
                    <OpprettEllerEndreNorskKontonr
                      personident={personident}
                      kontonummer={kontonr}
                    />
                  )}
                  <Radio
                    name={UTENLANDSK}
                    checked={fields.norskEllerUtenlandsk === UTENLANDSK}
                    label={msg({ id: "felter.kontonummervalg.utenlandsk" })}
                    onChange={(e) =>
                      setField({ norskEllerUtenlandsk: e.target.name })
                    }
                  />
                  {fields.norskEllerUtenlandsk === UTENLANDSK && (
                    <OpprettEllerEndreUtenlandsbank
                      personident={personident}
                      utenlandskbank={utenlandskbank}
                    />
                  )}
                  <div className="utbetalinger__knapper">
                    <div className="utbetalinger__knapp">
                      <Knapp
                        type={"standard"}
                        htmlType={"submit"}
                        disabled={submitted && !isValid}
                        autoDisableVedSpinner={true}
                        spinner={loading}
                      >
                        <FormattedMessage id={"side.lagre"} />
                      </Knapp>
                    </div>
                    <div className="utbetalinger__knapp">
                      <Knapp
                        type={"flat"}
                        htmlType={"button"}
                        disabled={loading}
                        onClick={() => settOpprettEllerEndre(false)}
                      >
                        <FormattedMessage id={"side.avbryt"} />
                      </Knapp>
                    </div>
                  </div>
                  {alert && <Alert {...alert} />}
                </RadioGruppe>
              );
            }}
          </Validation>
          <Kilde kilde="personalia.source.nav" lenkeType={"INGEN"} />
        </Form>
      ) : (
        <>
          {kontonr || utenlandskbank ? (
            <>
              <NorskKontonummer kontonummer={kontonr} />
              <Utenlandskonto utenlandskBankkonto={utenlandskbank} />
            </>
          ) : (
            <div className="underseksjon__beskrivelse">
              <FormattedMessage
                id="personalia.kontonr.ingenData"
                values={{
                  br: (text: String) => (
                    <>
                      <br />
                      {text}
                    </>
                  ),
                }}
              />
            </div>
          )}
          <Kilde
            kilde="personalia.source.nav"
            onClick={() => settOpprettEllerEndre(true)}
            lenkeTekst={
              kontonr || utenlandskbank ? "side.endre" : "side.leggtil"
            }
            lenkeType={"KNAPP"}
            ikon={kontonr || utenlandskbank ? endreIkon : leggTilIkon}
          />
        </>
      )}
    </Box>
  );
};

export default Utbetalinger;
