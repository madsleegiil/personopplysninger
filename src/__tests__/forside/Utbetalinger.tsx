import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { IntlProvider } from "react-intl";
import nbMessages from "text/nb";
import personInfo from "../../clients/apiMock/app/fetch/person-info.json";
import Utbetalinger from "../../pages/forside/sections/4-personinfo/4-utbetalinger/Utbetalinger";
import { StoreProvider } from "../../store/Context";

jest.mock("react-modal");

describe("Utbetalinger", () => {
  it("renders correctly", () => {
    const tree = renderer
      .create(
        <StoreProvider>
          <IntlProvider locale={"nb"} messages={nbMessages}>
            <Utbetalinger kontonr={personInfo.personalia.kontonr} />
          </IntlProvider>
        </StoreProvider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
