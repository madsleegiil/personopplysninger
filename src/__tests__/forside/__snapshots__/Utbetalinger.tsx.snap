// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Utbetalinger renders correctly 1`] = `
<div
  className="box__wrapper"
  id="utbetaling"
>
  <div
    className="nav-veilederpanel nav-veilederpanel--kompakt nav-veilederpanel--plakat"
  >
    <div
      className="nav-veileder nav-veileder--flytende"
    >
      <div
        className="nav-veileder__frame nav-veileder__frame--m"
      >
        <img
          alt="Veileder"
          className="box__ikon"
          src="Kontonummer.svg"
        />
      </div>
    </div>
    <div
      className="nav-veilederpanel__content"
    >
      <div
        className="box__container"
      >
        <div
          className="box__header"
        >
          <div
            className="box__title-container"
          >
            <div
              className="box__line"
            />
            <h2
              className="typo-systemtittel box__title"
            >
              Kontonummer for utbetaling
            </h2>
            <div
              className="box__line"
            />
          </div>
          <a
            className="lenke anchor-link"
            href="#utbetaling"
          >
            <img
              alt=""
              src="AnchorLink.svg"
            />
            Lenke hit
          </a>
        </div>
        <div
          className="box__content"
        >
          
          <ul
            className="list-column-2"
          >
            <li>
              <div
                className="listelement__header"
              >
                <p
                  className="typo-element"
                >
                  Kontonummer
                </p>
              </div>
              <div
                className="content "
              >
                1234 56 78911
              </div>
            </li>
          </ul>
          <div
            className="kilde__container"
          >
            <button
              className="kilde__lenke lenke"
              onClick={[Function]}
            >
              <span
                className="kilde__icon"
              >
                <img
                  alt="Ekstern lenke"
                  src="Pencil.svg"
                />
              </span>
              <p
                className="typo-normal"
              >
                Endre
              </p>
            </button>
            <div
              className="kilde__tekst"
            >
              <p
                className="typo-undertekst"
              >
                Kilde: NAV
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
